File: learn-spark/scala/base/src/main/scala/stream/TransformBlacklist.scala
package stream

import org.apache.spark.SparkConf
import org.apache.spark.streaming.{Seconds, StreamingContext}

object TransformBlacklist {

  def main(args: Array[String]): Unit = {
    val conf = new SparkConf()
      .setMaster("local[2]")
      .setAppName("TransformBlacklist")
    val ssc = new StreamingContext(conf, Seconds(5))

    val blacklist = Array(
      ("tom", true)
    )
    // 模拟一个黑名单
    val blacklistRDD = ssc.sparkContext.parallelize(blacklist, 5)

    // 输入数据格式为(date, name)
    val adsClickLogStream = ssc.socketTextStream("localhost", 9999)
    // 转换为 (name, (date, name)) 格式
    val userAdsClickLogStream = adsClickLogStream
      .map( adsClickLog => (adsClickLog.split(" ")(1), adsClickLog))

    val validAdsClickLogDStream = userAdsClickLogStream.transform( userAdsClickLogRDD => {

      // 转换为 (name, ((date, name), true)) 的格式
      val joinedRDD = userAdsClickLogRDD.leftOuterJoin(blacklistRDD)
      val filteredRDD = joinedRDD.filter(tuple => {
        if(tuple._2._2.getOrElse(false)){
          false
        } else {
          true
        }
      })

      val validAdsClickLogRDD = filteredRDD.map(tuple => tuple._2._1)
      validAdsClickLogRDD
    })

    validAdsClickLogDStream.print()
    ssc.start()
    ssc.awaitTermination()

  }
}

File: learn-spark/python/base_demo/streaming/transformBlacklist.py
#!/usr/bin/ python3
# -*- coding: utf-8 -*-
from pyspark.streaming import StreamingContext
from pyspark import SparkContext

sc = SparkContext("local[2]", "transformBlacklist")
ssc = StreamingContext(sc, 5)

blacklist = [
    ("tom", True),
    ("jerry", True)
]
# 模拟一个黑名单 (姓名, 是否启用)
blacklistRdd = sc.parallelize(blacklist)
# 接收实时访问日志, 这里简化为格式为 "date username
logDStream = ssc.socketTextStream("127.0.0.1", 9999)

# 把 "data username" 转为 (username, "date username)
userLogDStream = logDStream.map(lambda row: (row.split(" ")[1], row))


def transform(rdd):
    joinedRdd = rdd.leftOuterJoin(blacklistRdd)
    filteredRdd = joinedRdd.filter(lambda row: row[1][1] is not True)
    validLogRdd = filteredRdd.map(lambda row: row[1])
    return validLogRdd


validLogDStream = userLogDStream.transform(lambda rdd: transform(rdd))

validLogDStream.pprint()

ssc.start()
ssc.awaitTermination()
